### PR TITLE
feat(engine): Implement scope-aware scheduler + scatter-gather control flow

### PR DESCRIPTION
**Pull request by daryllimyt**
- **Fix DSLScheduler executor function type**
- **Introduce Task and ScopeID**
- **Completed tasks uses Task**
- **Set slots=True for DSLEdge**
- **pWIP] First pass**
- **Replace refs with task instances in success/fail path handlers**
- **Add scope ID contextvar**
- **Add cooperate for loop**
- **Add test**
- **Working nested ex/im**
- **Add more tests**
- **Pass stmt and task in fn args + make run_if scope-aware**
- **Add implode unset sentinel**
- **Tenacity retrying -> being action attempt**
- **Use Required and NotRequired instead of total=False**
- **Working filtering**
- **Remove drop_unset**
- **Add registry actions**
- **Accept objects in explode collection**
- **Scope -> stream**
- **Got working empty collection edge case handling**
- **Add 1D and 2D explode with multiple parallel if conditions inside**
- **Tests for parallel blocks**
- **Rename explode -> scatter, implode -> gather**
- **Update action description and paramter docs**
- **Rename to scatter-gather 2**
- **Mixed collection tests**
- **Rename**
- **Cleanup traceback**
- **Remove scope_id in Task and DSLEdge + use root stream instead of None**
- **Remove counter**
- **Docstrings on Task**
- **Update test names + add skip all iterations**
- **Add error handling strategy**
- **Make TaskResult generic**
- **Make scheduler logic more uniform again. No special handling for root skip streams. Add err handling strategies**
- **Fix is error info logic**
- **Remove lru cache as it doesn't work with some types**
- **Add tests checking out-of-stream lookups**
- **Cleanup and manage log levels**
- **Remove tb**
- **Simplify while loop**
- **Drop ScopeBoundary**
- **Drop DSLScheduler class attr**
- **Add dsl scope validation tests**
- **Add dsl scope validation**
- **Tidy up**
- **Allow refs from inner to outer scopes**
- **Cleanup massive prints**
- **Tune logging**

### CODE REVIEW

File: tracecat/dsl/common.py [hunk 1]
```diff
@@ -126,29 +155,221 @@ def validate_structure(self) -> Self:
...
+                            f"Scatter action {action.ref!r} depends on {dep!r}, which isn't the parent scope"
+                        )
+                elif action.action == PlatformAction.TRANSFORM_GATHER:
+                    # Here, action_scope is the parent scope
+                    if action_scope != scope_hierarchy[dep_scope]:
+                        raise TracecatDSLError(
+                            f"Gather action {action.ref!r} depends on {dep!r}, which isn't the child scope"
+                        )
+                else:
+                    # Dep scope must be the same as action_scope or an ancestor (parent, grandparent, etc.)
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tracecat/dsl/common.py]: This is expected

File: tracecat/dsl/workflow.py [hunk 2]
```diff
@@ -903,13 +917,23 @@ async def _prepare_child_workflow(self, task: ActionStatement) -> DSLRunArgs:
         )
 
     async def _run_action(self, task: ActionStatement) -> Any:
+        # XXX(perf): We shouldn't pass the full execution context to the activity
+        # We should only keep the contexts that are needed for the action
+        stream_id = CTX_STREAM_ID.get()
+        expr_ctxs = extract_expressions(task.model_dump())
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tracecat/dsl/workflow.py:957]: its a pydantic model so the structure is valid

File: tracecat/dsl/models.py [hunk 3]
```diff
@@ -26,16 +26,16 @@
 """Workflow execution context."""
 
 
-class TaskResult(TypedDict, total=False):
+class TaskResult[TResult: Any, TError: Any](TypedDict):
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tracecat/dsl/models.py:37]: we use python 3.12

File: registry/tracecat_registry/core/transform.py [hunk 4]
```diff
@@ -195,3 +195,46 @@ def compact(
...
+    default_title="Scatter",
+    description=(
+        "Transform a collection of items into parallel execution streams, "
+        "where each item is processed independently."
+    ),
+    display_group="Data Transform",
+    namespace="core.transform",
+)
+def scatter(
+    collection: Annotated[
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply registry/tracecat_registry/core/transform.py:210]: We handle this using `ScatterArgs`

File: tests/unit/test_workflows.py [hunk 5]
```diff
@@ -3181,3 +3187,1603 @@ async def test_workflow_detached_child_workflow(
...
+                        args=GatherArgs(
+                            items="${{ ACTIONS.identity.result }}"
+                        ).model_dump(),
+                    ),
+                ],
+            ),
+            {
+                "ACTIONS": {
+                    "gather": {
+                        "result": [1, "hello", {"key": "value"}, None, [1, 2]],
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tests/unit/test_workflows.py:3874]: Expected

File: tests/unit/test_workflows.py [hunk 6]
```diff
@@ -3181,3 +3187,1603 @@ async def test_workflow_detached_child_workflow(
...
+            ),
+            {
+                "ACTIONS": {
+                    "gather1": {
+                        "result": [],
+                        "result_typename": "list",
+                        "error": [
+                            {
+                                "ref": "throw",
+                                "message": "There was an error in the executor when calling action 'core.transform.reshape'.\n\n\nTracecatExpressionError: Error evaluating expression `1/0`\n\n[evaluator] Evaluation failed at node:\n```\nbinary_op\n  literal\t1\n  /\n  literal\t0\n\n```\nReason: Error trying to process rule \"binary_op\":\n\nCannot divide by zero\n\n\n------------------------------\nFile: /app/tracecat/expressions/core.py\nFunction: result\nLine: 73",
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tests/unit/test_workflows.py:4509]: Ignore for now

File: tests/unit/test_workflows.py [hunk 7]
```diff
@@ -3181,3 +3187,1603 @@ async def test_workflow_detached_child_workflow(
...
+                        args={"value": "${{ 1/0 }}"},
+                    ),
+                    # This gather collects the results from the outer scatter1.
+                    ActionStatement(
+                        ref="gather1",
+                        action="core.transform.gather",
+                        depends_on=["throw"],
+                        args=GatherArgs(
+                            items="${{ ACTIONS.throw.result }}",
+                            error_handling_strategy=StreamErrorHandlingStrategy.INCLUDE,
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tests/unit/test_workflows.py]: Self explanatory

File: tests/unit/test_dsl_scope_validation.py [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+    """Test scope region validation for scatter-gather patterns."""
+
+    def test_valid_scatter_gather_workflow(self):
+        """Test a valid scatter-gather workflow passes validation."""
+        dsl = DSLInput(
+            title="Valid Scatter-Gather",
+            description="A workflow with proper scatter-gather scoping",
+            entrypoint=DSLEntrypoint(),
+            actions=[
+                ActionStatement(  # scope: root
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tests/unit/test_dsl_scope_validation.py:20]: intentional

File: tracecat/dsl/workflow.py [hunk 9]
```diff
@@ -600,7 +613,8 @@ async def _execute_task(self, task: ActionStatement) -> TaskResult:
             raise ApplicationError(msg, non_retryable=True, type=err_type) from e
         finally:
             logger.trace("Setting action result", task_result=task_result)
-            self.context[ExprContext.ACTIONS][task.ref] = task_result  # type: ignore
+            context = self.get_context(stream_id)
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tracecat/dsl/workflow.py:618]: We have workflow-level error handling which would catch this

File: tracecat/dsl/workflow.py [hunk 10]
```diff
@@ -903,13 +917,23 @@ async def _prepare_child_workflow(self, task: ActionStatement) -> DSLRunArgs:
         )
 
     async def _run_action(self, task: ActionStatement) -> Any:
+        # XXX(perf): We shouldn't pass the full execution context to the activity
+        # We should only keep the contexts that are needed for the action
+        stream_id = CTX_STREAM_ID.get()
+        expr_ctxs = extract_expressions(task.model_dump())
+        new_action_context: dict[str, Any] = {}
+        for action_ref in expr_ctxs[ExprContext.ACTIONS]:
+            res = self.scheduler.get_stream_aware_action_result(action_ref, stream_id)
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tracecat/dsl/workflow.py:960]: Logging this would be too verbose

File: tracecat/dsl/workflow.py [hunk 11]
```diff
@@ -903,13 +917,23 @@ async def _prepare_child_workflow(self, task: ActionStatement) -> DSLRunArgs:
         )
 
     async def _run_action(self, task: ActionStatement) -> Any:
+        # XXX(perf): We shouldn't pass the full execution context to the activity
+        # We should only keep the contexts that are needed for the action
+        stream_id = CTX_STREAM_ID.get()
+        expr_ctxs = extract_expressions(task.model_dump())
+        new_action_context: dict[str, Any] = {}
+        for action_ref in expr_ctxs[ExprContext.ACTIONS]:
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tracecat/dsl/workflow.py:959]: Fine IMO, platform level issue if it fails

File: tracecat/workflow/executions/service.py [hunk 12]
```diff
@@ -221,10 +221,11 @@ async def list_workflow_execution_events_compact(
                     source.action_error = EventFailure.from_history_event(event)
         # For the resultant values, if there are duplicate source action_refs,
         #  we should merge them into a single event.
-        finalRef2events: dict[str, WorkflowExecutionEventCompact] = {}
+        task2events: dict[Task, WorkflowExecutionEventCompact] = {}
         for event in id2event.values():
-            if event.action_ref in finalRef2events:
-                group_event = finalRef2events[event.action_ref]
+            task = Task(ref=event.action_ref, stream_id=event.stream_id)
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tracecat/workflow/executions/service.py:226]: Stream ID always exists on event

File: tracecat/workflow/executions/service.py [hunk 13]
```diff
@@ -221,10 +221,11 @@
                     source.action_error = EventFailure.from_history_event(event)
         # For the resultant values, if there are duplicate source action_refs,
         #  we should merge them into a single event.
-        finalRef2events: dict[str, WorkflowExecutionEventCompact] = {}
+        task2events: dict[Task, WorkflowExecutionEventCompact] = {}
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tracecat/workflow/executions/service.py:224]: Frozen dataclass automatically generates __hash__()

File: tracecat/dsl/models.py [hunk 14]
```diff
@@ -26,16 +34,16 @@
...
```
[the remark this answers is not in the document]
Comment (daryllimyt) [reply tracecat/dsl/models.py:37]: This works in python 3.12
